build_runner_core/lib/src/package_graph/target_graph.dart

```diff
@@ -82,8 +83,14 @@ class TargetGraph {
         publicAssetsByPackage[package.name] =
             InputMatcher(const InputSet(), defaultInclude: defaultInclude);
       }
-      final nodes = config.buildTargets.values.map((target) =>
-          TargetNode(target, package, defaultInclude: defaultInclude));
+      final nodes = config.buildTargets.values.map((target) {
+        if (target.sources.includeDefaults &&
+            !sourcesWithIncludedDefaults.contains(target.sources)) {
```

Comment:
I am not quite understanding the point of this `sourcesWithIncludedDefaults` set, can you explain?

If the issue is a shared `sources` list across targets, then I think we have a different issue here.

Reply:
@jakemac53 When I tested my code the first time, I found that a map function called several times. I didn't do a deep digging to find why, but now, when you have pointed to it I have done and I found, that a `nodes` variable has `Iterable` type, so, when we try to get values from the `nodes` variable in lines https://github.com/dart-lang/build/blob/1dc5aa685909a126ae2ef6e527bf392b603b8ac4/build_runner_core/lib/src/package_graph/target_graph.dart#L212 and https://github.com/dart-lang/build/blob/1dc5aa685909a126ae2ef6e527bf392b603b8ac4/build_runner_core/lib/src/package_graph/target_graph.dart#L108
We call the map function again. 

So, we can easily fix this issue just if we call `toList` method after `map` but, maybe you know another reason why we need to create a new `TargetNode` instance in the map function again and again?

Reply:
@jakemac53 I don't want to add `sources` in `InputSet` directly as well. But if we don't do it, we will have a problem with graphs as the graph doesn't use `InputMatcher`. For example here: 
https://github.com/dart-lang/build/blob/d2b84fd34ca6825ced31f5a0ecf3b38bbb5747d2/build_runner_core/lib/src/package_graph/apply_builders.dart#L346

We can fix it by changing `node.target.sources` to something like that: 
```dart
return createPhase(
  node.package,
  options,
  InputSet(
      include: node.sourceIncludes.map((e) => e.toString()).toList()),
  builderConfig?.generateFor,
  isReleaseMode);
```

But I'm not sure that my changes don't break something else. @jakemac53 how do you think are these changes worth it? If it is, so, we need to spend more time for testing graphs and etc as well.

Reply:
Ultimately, we always convert the `InputSet` to an `InputMatcher` before using it, so I don't see a real issue here with passing `node.target.sources` directly there, it is immediately converted to an `InputMatcher` though so we could also more eagerly convert it.

It is interesting though how the `InBuildPhase` does not pass a `defaultInclude` for the `InputMatcher` that it creates. I think we could probably change `InputMatcher` to require the `defaultInclude` parameter though, to flag all places where it isn't passed (looks like only InBuildPhase), and then pass in the desired defaults there, which would require some additional plumbing. Possibly the `InBuildPhase` would need a `PackageNode` instead of just a `String` with the package name (but it looks like this is available, although it would break a bunch of tests).

Reply:
But note that it has been a very long time since I looked at any of this code so there is definitely some investigation required here.